storybookjs/storybook
# Addon-controls: Add JSON tree editor for Object/Array Type args

Issue: #12599 , reimplements https://github.com/storybookjs/storybook/pull/12685

## What I did

- Replaced the existing text form control with the `react-editable-json` lib (public demo here: https://oxyno-zeta.github.io/react-editable-json-tree/ ). I also added typings for the library to the closest typings file so we can safely use it in typescript.
- Made a first pass at customizing the styles (spacing and colors) based on @domyen's feedback [here](https://github.com/storybookjs/storybook/pull/12685#issuecomment-705017673) and looking at the colors available in Storybook [here](https://next--storybookjs.netlify.app/official-storybook/?path=/story/basics-colorpalette--page) . The "dark/light mode toggles" aren't toggling the styles in the control pane (just inside the main component body area), so I'm not exactly how this will look in Dark Mode yet, but I set up the style customization function such that it should be easy to replace things
- Updated the `official-storybook` story about using `addon-controls` to make use of the fact that we now handle nested objects.

- Demo GIF: https://a.cl.ly/geuqpZrq

## How to test

- Is this testable with Jest or Chromatic screenshots?

It's possible, but I'm not sure that it's necessary.

- Does this need a new example in the kitchen sink apps?

We could, but I think updating the main storybook story may be sufficient, as this change should benefit most users without needing them to customize anything. 

- Does this need an update to the documentation?

We could update [this page](https://storybook.js.org/docs/react/essentials/controls#annotation) to indicate that tree view is used instead of a text view.

## FAQ

- Why not some other JSON tree viewer library XYZ?
  - We explored several different alternatives, tradeoffs are documented [here](https://github.com/storybookjs/storybook/pull/12685#issuecomment-708699165), [here](https://github.com/storybookjs/storybook/pull/12685#issuecomment-706739298), and [here](https://github.com/storybookjs/storybook/issues/12599#issuecomment-701097502) 

- What terminal commands did you use that you might use again next time?

```bash
# to add a dependency scoped to 1 package 
yarn lerna add react-editable-json-tree --scope=@storybook/components

# to test locally
yarn build components --watch
yarn start
```

- Any modifications useful in dev that you didn't commit?

 1 change in `examples/official-storybook/main.ts` to speed up build time:

```js
{
stories: [
    './stories/**/addon-controls.stories.@(js|ts|tsx|mdx)',
  ]
 // ... rest
}
```
